### ISSUE TITLE
Integrating deep q learning algo written in python with MATSIM 

### ISSUE THREAD

**Issue by ashu170292** [matsim-org/matsim-code-examples#84]
I am trying to develop a distributed model free algorithm for ride sharing using deep reinforcement learning. I want to create a simulator design that uses Deep Q techniques to learn optimal dispatch policies for taxis through interaction with environment. Since I have used python to develop the codes for reinforcement learning and am unable to find a way to integrate my algorithm with MATSIM because there is no python API, can someone please guide me into how this can be achieved

**Comment by michalmac** [comment]
Unfortunately, there is no such API. Please have a look at #34 for a general discussion, and the solution proposed by @tduberne .

**Comment by jfbischoff** [comment]
Perhaps the framework proposed as part of the AI driving challenge might also be of interest:
https://matsim.atlassian.net/wiki/spaces/MATPUB/blog/2018/12/05/351240193/AI+Driving+Olympics+Managing+an+Automated+Taxi+Fleet

**Comment by tduberne** [comment]
As mentioned by @michalmac, you might want to have a look at https://github.com/tduberne/matsim-python-experiments

As far as I am concerned, this fulfills all requirements I have for such an API:
- I can import and manipulate any MATSim class as if it was a Python class, for instance for converting a set of Pandas DataFrames to MATSim containers and writing them to files, or feeding them directly to a Controler and starting it.
- I can even implement Java interfaces in Python, for instance writing an event-handler that stores events in a Pandas DataFrame.

The first usecase is basically what I would expect under "Python API for MATSim", so I actually consider that API to exist already. Here performance is not much of an issue: the "heavy lifting" happens in pure Java.

The second one is a nice toy and can be nice in teaching, to avoid having to teach two languages for students, but it has several drawbacks over doing it in Java:
- communication between Java and Python comes with some overhead, so it can slow things down quite a bit if there are lots of calls to your Python classes (as in the case of event handling). But if I trust them, it is as good as one can get.
- IDE support is much, much better in Java. Actually, due to the way JPype works, you do not get _any_ autocompletion support. MATSim's "user experience" depends a lot on IDE support, so doing anything complex in Python requires significantly more brainpower.

If I understand correctly, what you want is more in line with that second usecase. It should work, and might be fast enough if you reduce communication between Java and Python to a minimum. That would probably mean that you need to implement some infrastructure in Java that delegates the heavy computations to some interface, and that you (can) implement this interface in Python. But if you want to do complicated things, you will have to use a JVM-based language. And even if going the Python way, you would benefit from first experimenting in Java, to understand a bit better how things work in the java side before replacing the guts with python, if needed.

*TL;DR*
If you really need to keep your stuff in Python, JPype is your best bet.
But if possible, I would strongly advise looking at JVM-based options (Java, Scala...).
